crates/sui-framework/sources/test_scenario.move.WIP

```diff
@@ -0,0 +1,363 @@
+        let id_opt = most_recent_id_for_address<T>(account);
+        assert!(option::is_some(&id_opt), EEmptyInventory);
+        take_from_address_by_id(account, option::destroy_some(id_opt))
+    }
+
+    /// Return `t` to the inventory of the `account`
+    public fun return_to_address<T: key>(account: address, t: T) {
+        let id = object::id(&t);
+        assert!(was_taken_from_address(account, id), ECantReturnObject);
+        sui::transfer::transfer(t, account)
+    }
+
+    /// Returns true if the object with `ID` id was in the inventory for `account`
+    public native fun was_taken_from_address(account: address, id: ID): bool;
+
+    // == from sender ==
+
+    /// helper for `take_from_address_by_id` that operates over the transaction sender
+    public fun take_from_sender_by_id<T: key>(scenario: &Scenario, id: ID): T {
+        take_from_address_by_id(sender(scenario), id)
+    }
+
+    /// helper for `most_recent_id_for_address` that operates over the transaction sender
+    public fun most_recent_id_for_sender<T: key>(scenario: &Scenario): Option<ID> {
+        most_recent_id_for_address<T>(sender(scenario))
+    }
+
+    /// helper for `take_from_address` that operates over the transaction sender
+    public fun take_from_sender<T: key>(scenario: &Scenario): T {
+        take_from_address(sender(scenario))
+    }
+
+    /// helper for `return_to_address` that operates over the transaction sender
+    public fun return_to_sender<T: key>(scenario: &Scenario, t: T) {
+        return_to_address(sender(scenario), t)
+    }
+
+    /// Returns true if the object with `ID` id was in the inventory for the sender
+    public fun was_taken_from_sender(scenario: &Scenario, id: ID): bool {
+        was_taken_from_address(sender(scenario), id)
+    }
+
+    // == immutable ==
+
+    /// Remove the immutable object of type `T` with ID `id` from the global inventory
+    /// Aborts if there is no object of type `T` in the inventory with ID `id`
+    public native fun take_immutable_by_id<T: key>(id: ID): T;
+
+    /// Returns the most recent immutable object of type `T`
+    public native fun most_recent_immutable_id<T: key>(): Option<ID>;
+
+    /// Helper combining `take_immutable_by_id` and `most_recent_immutable_id`
+    /// Aborts if there is no immutable object of type `T` in the global inventory
+    public fun take_immutable<T: key>(): T {
+        let id_opt = most_recent_immutable_id<T>();
+        assert!(option::is_some(&id_opt), EEmptyInventory);
+        take_immutable_by_id(option::destroy_some(id_opt))
+    }
+
+    /// Return `t` to the global invetory
+    public fun return_immutable<T: key>(t: T) {
+        let id = object::id(&t);
+        assert!(was_taken_immutable(id), ECantReturnObject);
+        sui::transfer::freeze_object(t)
+    }
+
+    /// Returns true if the object with `ID` id was an immutable object in the global invetory
+    public native fun was_taken_immutable(id: ID): bool;
```

Comment:
Does this need to be `public` (and similar to `was_taken_shared`)? Trying to imagine when a tester would use these.

Reply:
Let me look again

Reply:
We can make them private for now, can always change it later